### ISSUE TITLE
mv: semaphore units are not kept alive until a view update finishes

### ISSUE THREAD

**Issue by wmitros** [scylladb/scylladb#17890]
When performing a view update, we take some units from view_update_concurrency_semaphore to track memory used by it:
https://github.com/scylladb/scylladb/blob/e48eb76f61046d27af5801a5a52e758ab40d08c8/replica/table.cc#L2650
These units should be in scope for as long as we're using the resources, however, in a case where the update needs to be performed both locally and remotely, the units are only held until the local update finishes:
https://github.com/scylladb/scylladb/blob/e48eb76f61046d27af5801a5a52e758ab40d08c8/db/view/view.cc#L1776-L1778
and no semaphore units are left for the remote update:
https://github.com/scylladb/scylladb/blob/e48eb76f61046d27af5801a5a52e758ab40d08c8/db/view/view.cc#L1812-L1814
Instead, in the described case all the semaphore units should be kept alive until the remote update finishes.

Due to releasing semaphore units while holding on to the resources we can allocate even more resources, eventually leading to OOM.

This problem only occurs when we're performing both local and remote changes for a view update, which is our behavior during node operations, particularly when a node is joining. Potentially, this may also happen during tablet operations, when they're enabled.

**Comment by nyh** [comment]
Since the remote view update happens (usually) asynchronously, we don't wait for it and release the semaphore units before the remote update completes, doesn't this completely break the idea that the semaphore counts the amount of memory used by pending view updates??? Is it possible we had a bug where we only actually counted local view updates and never counted remote view updates, and we were just lucky that if for example 1/3rd of the updates are local (by chance), counting 1/3rd of the pending updates was good enough?

How did you see this bug in practice?

It's really sad how we had so many bugs in this code and nobody ever noticed any of them because we never actually tested any of this code with any degree of rigorousness. Thanks @wmitros for doing this.

**Comment by avikivity** [comment]
Please adjust the description to include a failure scenario, with impact, to allow us to understand whether it needs to be backported.

**Comment by wmitros** [comment]
> Since the remote view update happens (usually) asynchronously, we don't wait for it and release the semaphore units before the remote update completes, doesn't this completely break the idea that the semaphore counts the amount of memory used by pending view updates??? Is it possible we had a bug where we only actually counted local view updates and never counted remote view updates, and we were just lucky that if for example 1/3rd of the updates are local (by chance), counting 1/3rd of the pending updates was good enough?

From the bugs I'm seeing (https://github.com/scylladb/scylladb/issues/17364 and https://github.com/scylladb/scylladb/issues/17854), it doesn't look like what we were counting ever really mattered (the `view_update_concurrency_semaphore` was also just for tracking). It's surprising that we didn't have even more issues due to this.

> How did you see this bug in practice?

I encountered it while comparing the approaches for throttling large view updates. When running the approach with waiting on a semaphore I was getting OOMs while it should have been the surest way of preventing them. It may also explain some OOMs I encountered while testing the sleep-based approach which I attributed to our lack of admission control

> It's really sad how we had so many bugs in this code and nobody ever noticed any of them because we never actually tested any of this code with any degree of rigorousness. Thanks @wmitros for doing this.

> Please adjust the description to include a failure scenario, with impact, to allow us to understand whether it needs to be backported.

As long as we're backporting https://github.com/scylladb/scylladb/issues/17364 and https://github.com/scylladb/scylladb/issues/17854, I'd assume we want to backport this as well. The scenario we're trying to help here is the same as in the other 2 issues: reaching OOM in MV-heavy workloads (particularly ones with partition deletes).

**Comment by wmitros** [comment]
> Please adjust the description to include a failure scenario, with impact, to allow us to understand whether it needs to be backported.

I adjusted the description again, I noticed the issue is not that common - we're only performing both local and remote updates during node operations. Because they're rather rare, this issue is not that likely to happen. It may be a bigger problem when we enable tablets, however.

**Comment by nyh** [comment]
> This problem only occurs when we're performing both local and remote changes for a view update, which is our behavior during node operations, particularly when a node is joining. This may also potentially also happen during tablet operations, when they're enabled.

Interesting. I forgot this code path (a single iteration in the parallelized loop of mutate_MV()) only handles a single view update from a single base replica - so normally it only does one update - which is either local or remote but not both.
And if I understand correctly, the old code was already correct (albeit, hard to reason about) in both of the common cases (remote-only and local-only).

So now I'm puzzled: How did you ever see this bug in practice? You said you got OOMs in your test case, but how come your test case had this problem when you say now it only happens during range movements? Did you have range movements during your test?

**Comment by wmitros** [comment]
> > This problem only occurs when we're performing both local and remote changes for a view update, which is our behavior during node operations, particularly when a node is joining. This may also potentially also happen during tablet operations, when they're enabled.
> 
> Interesting. I forgot this code path (a single iteration in the parallelized loop of mutate_MV()) only handles a single view update from a single base replica - so normally it only does one update - which is either local or remote but not both. And if I understand correctly, the old code was already correct (albeit, hard to reason about) in both of the common cases (remote-only and local-only).
> 
> So now I'm puzzled: How did you ever see this bug in practice? You said you got OOMs in your test case, but how come your test case had this problem when you say now it only happens during range movements? Did you have range movements during your test?

While testing the range movements and view updates together wasn't my intention, I did need to run tests on a (2-node) cluster (to have remote updates) and that run must have started before the second node joined.

**Comment by wmitros** [comment]
> While testing the range movements and view updates together wasn't my intention, I did need to run tests on a (2-node) cluster (to have remote updates) and that run must have started before the second node joined.

This was also the scenario I was testing the solution on - I was getting OOMs before and wasn't afterwards. It would probably be useful to also test node operations after initializing the cluster, but it's just one of missing MV tests that we would like to have

**Comment by nyh** [comment]
> While testing the range movements and view updates together wasn't my intention, I did need to run tests on a (2-node) cluster (to have remote updates) and that run must have started before the second node joined.

Was it a 2-node cluster with RF=1? Yes, I guess that does involve range movements.
Nice to have a test for this case! (Even if it's not really what you intended to test)
